Added extension hook for reading modes

kevinpapst opened the pull request:
I am currently working on a full-fledged extension, which uses its own controllers and views.

Problem is: I cannot route the user to the page (controller/action) that I want him to see as there is currently no way of displaying additional links in the main UI.

To achieve my goal, I refactored the reading modes to be a simple array structure which gets injected into the hook. Extensions can then work on that navigation structure like that:

```
class MyExtension extends Minz_Extension
{
    public function init() {
	$this->registerHook('nav_reading_modes', array($this, 'addReadingModeToMenu'));
    }

    public function addReadingModeToMenu($menu) {
    	$request = Minz_Request::currentRequest();
	$menu[] = new FreshRSS_ReadingMode(
		'example',
		'example title',
		array_merge($request, array('c' => 'my', 'a' => 'any')),
		($request['c'] == 'my' && $request['a'] == 'any')
	);
	return $menu;
    }
}
```

Now extensions are capable of adding their own reading modes into the menu.

What do you think of this approach?

Frenzie commented:
Should all of that `$readingModes` definition stuff really go in the template? That looks like template pollution to me. But in principle I think it's clearer this way even without extensions.

kevinpapst commented:
True, its not the best place to keep that logic, but I was unsure where to put it otherwise. Any idea? We could create a base `FreshRSS_Controller` which extends `Minz_Controller` and then let the `FreshRSS_index_Controller` extend that new `FreshRSS_Controller`. Then extensions can extend this controller as well if they want to display the `nav_menu.phtml` partial.

Alkarex commented:
I have not looked in details yet, but could this code be in `FreshRSS_index_Controller` and you could override `FreshRSS_index_Controller` when needed?

Alkarex commented on app/layout/nav_menu.phtml:
```diff
@@ -131,20 +131,38 @@
+				'name' => _i("view-normal"),
+				'title' => _t('index.menu.normal_view'),
+				'url' => array_merge($url_output, array('c' => Minz_Request::defaultControllerName(), 'a' => 'normal')),
+				'active' => (Minz_Request::controllerName() == Minz_Request::defaultControllerName() && $actual_view == 'normal'),
```
Please put the "fast" string comparisons first, and use triple-equal, like:

```php
'active' => $actual_view === 'normal' && Minz_Request::controllerName() === Minz_Request::defaultControllerName(),
```

and maybe the multiple `Minz_Request::controllerName() === Minz_Request::defaultControllerName()` could be factorised somehow

kevinpapst commented:
Updated the PR: refactored ReadingMode to a Model class. Keeping it in IndexController would force Extensions to extend that IndexController. But I could still move it there, if you disagree with the model.

Alkarex commented:
Sorry for the delay.
Happy New Year 2018 :-)

Frenzie commented on app/Models/ReadingMode.php:
```diff
@@ -0,0 +1,136 @@
+
+		$readingModes = array(
+			new FreshRSS_ReadingMode(
+				_i("view-normal"),
```
What if we just make the argument here `"view-normal"` and then above something like

```php
$this->id = $name;
$this->name = _i($name);
```

This is in reference to #1758.

Frenzie commented on app/layout/nav_menu.phtml:
```diff
@@ -131,20 +131,19 @@
+		/** @var FreshRSS_ReadingMode $mode */
+		foreach ($readingModes as $mode) {
+			?>
+			<a class="view_normal btn <?php if ($mode->isActive()) { echo 'active'; } ?>" title="<?php echo $mode->getTitle(); ?>" href="<?php echo Minz_Url::display($mode->getUrlParams()); ?>">
```
And then the ID would be part of the className here. Sorry for kind of polluting your PR but I figure it's easier to discuss this way than in #1758.

aledeg replied:
That was the idea behind the class ```view_normal```, ```view_global```, and ```view_reading```

kevinpapst commented:
I did not follow the changes in #1755. From a quick view I would say its slightly incompatible with my changes? 
Now extension can add views, which then would not have a shortcut. And the list of views could even be completely rewritten by the event, so the hard coded shortcuts in #1758 may be completely wrong.
I can add your changes @Frenzie then we have view IDs which can be hardwired to the shortcuts. In the unrealistic case that a view link is not existing in the list/DOM the JS would not trigger.
Is that what you suggested @Frenzie ? @aledeg any objections?

Frenzie commented:
@kevinpapst I'm not entirely sure what you're suggesting with hardwiring but I suppose you mean what I mean, lol. I was suggesting to dynamically restore the HTML/DOM to the way it was or at least something close to it. That way a selector like  `#nav_menu_views .view-rss` would work, although I mean that mostly for styling purposes.

I'm not the biggest fan of the JS using the DOM like in #1755 since my DOM might be different, but what I thought reasonable enough with classes could yield odd and unexpected results while using the *n*th child node. (Including not working, but that's pretty much the best case scenario.) Ideally the info from `Minz_Request::currentRequest()` would probably also be easily available to the JS. But that aside.

PS Incidentally, [here](https://stackoverflow.com/a/11654596/2470572) is the first search result for a premade function that modifies a single parameter in the query string, although just the value of `a=blabla` would likely do. Anyway, that's just a quick brainstorm.

aledeg commented:
The problem is the class used. Before, each view had a different CSS class (```view_normal```, ```view_global```, ```view_reading```)(See https://github.com/FreshRSS/FreshRSS/pull/1755/files#diff-99959fbf8401f92306d451127be5e01a). Now, they all have the ```view_normal``` class.
I am not saying that my fix was perfect. But it fixes the issue at the moment.

What should be done is restoring a way to have a different CSS class for each of the buttons. This way we can revert the fix to its initial setting.

kevinpapst commented:
Ok, now I am slightly confused^^ Let me adjust the code and submit a PR hopefully tomorrow.

I am concerned about the changes in #1758 as they can fail in rare circumstances. The hardcoded index used in the get() call can be wrong if an extension manipulates the available reading modes in any way. So we are concerned about the same point but probably for different reasons ;-)
Therefor we need to re-add the class or an index to the links, which were previously available and which I removed because they were just not used before @aledeg introduced his changes.

aledeg commented:
No hard feelings, we all make mistakes. I've recently rewrote a PR after Alkarex mentioned something I didn't see :)

kevinpapst commented:
Hahaha, I blame everything on you @aledeg ... nobody would have noticed without your changes LOL
I ping you with my fixes tomorrow for review.

Frenzie commented:
> What should be done is restoring a way to have a different CSS class for each of the buttons. This way we can revert the fix to its initial setting.

@kevinpapst This is what I meant to convey in my comments, btw. I figured the translation ID could also serve well doing double duty as a className.

aledeg commented:
@kevinpapst @Frenzie I think we were all taking about the same thing :)

aledeg commented:
@kevinpapst any news regarding the class fix?

kevinpapst commented:
Ups, totally forgot about it, thanks for the reminder. I'am on it right now

kevinpapst commented:
@aledeg see #1773
@Frenzie thanks, just did it as you proposed

aledeg commented:
@kevinpapst great!
